File: http-streams/src/main/java/streams/Intro.java
package streams;

import akka.Done;
import akka.NotUsed;
import akka.actor.ActorSystem;
import akka.stream.ActorMaterializer;
import akka.stream.Materializer;
import akka.stream.javadsl.Flow;
import akka.stream.javadsl.RunnableGraph;
import akka.stream.javadsl.Sink;
import akka.stream.javadsl.Source;

import java.util.concurrent.CompletionStage;

public class Intro {
  void run() {
    //#source
    Source<Integer, NotUsed> source =
      //#source-no
      Source.range(0, 20000000);
      //#source-no
    //#source
    //#flow
    Flow<Integer, String, NotUsed> flow =
      //#flow-no
      Flow.fromFunction(n -> n.toString());
      //#flow-no
    //#flow
    //#sink
    Sink<String, CompletionStage<Done>> sink =
      //#sink-no
      Sink.foreach(str -> System.out.println(str));
      //#sink-no
    //#sink
    //#graph
    RunnableGraph<NotUsed> runnable =
      source.via(flow).to(sink);
    //#graph
    //#run
    ActorSystem system = ActorSystem.create();
    Materializer materializer = ActorMaterializer.create(system);

    runnable.run(materializer);
    //#run

    //#short
    Source.range(0, 20000000)
      .map(Object::toString)
      .runForeach(str -> System.out.println(str), materializer);
    //#short
  }
}

File: http-streams/src/main/scala/streams/Intro.scala
package sc.streams

import akka.actor.ActorSystem
import akka.stream.ActorMaterializer
import akka.stream.scaladsl.{Flow, Sink, Source}

class Intro {
  //#source
  val source = Source(0 to 20000000)
  //#source
  //#flow
  val flow = Flow[Int].map(_.toString())
  //#flow
  //#sink
  val sink = Sink.foreach[String](println(_))
  //#sink
  //#graph
  val runnable = source.via(flow).to(sink)
  //#graph
  //#run
  implicit val system = ActorSystem()
  implicit val mat = ActorMaterializer()

  runnable.run()
  //#run

  //#short
  Source(0 to 20000000)
    .map(_.toString)
    .runForeach(println)
  //#short
}
